# Rule ENT001

## Definition

Prefer JSON notation array format for CMD and ENTRYPOINT

## Description



## Examples


 &#x1F7E2; &nbsp; Proper ENTRYPOINT command in exec JSON format.

```Dockerfile
FROM golang 1.16
ENTRYPOINT [&amp;#34;/bin/bash&amp;#34;, &amp;#34;date&amp;#34;]
```


 &#x1F534; &nbsp; Proper ENTRYPOINT command in shell format.

```Dockerfile
FROM golang 1.16
ENTRYPOINT /bin/bash date
```



<details><br>
<summary>Additional examples</summary>


 &#x1F534; &nbsp; Proper ENTRYPOINT command in invalid format with 2 args.

```Dockerfile
    FROM golang 1.16
    ENTRYPOINT [/bin/bash date]
```


 &#x1F534; &nbsp; Proper ENTRYPOINT command in shell format.

```Dockerfile
    FROM golang 1.16
    ENTRYPOINT date
```


 &#x1F534; &nbsp; Proper ENTRYPOINT command in invalid format with 1 arg.

```Dockerfile
    FROM golang 1.16
    ENTRYPOINT [date]
```


 &#x1F534; &nbsp; Proper ENTRYPOINT command in exec JSON format, but missing a comma.

```Dockerfile
    FROM golang 1.16
    ENTRYPOINT [&amp;#34;/bin/bash&amp;#34; &amp;#34;date&amp;#34;
```



<p align="right"><sup>Note: all examples are parsed and/or generated from test cases.</sup></p>

</details>

## Reference

- TODO
